README
HTML::Lint Version 0.90

HTML::Lint is broken down into a number of perl modules:

	HTML::Lint		- provides the "weblint engine"
	HTML::Lint::HTML4	- encapsulation of messages
	HTML::Lint::Messages	- encapsulation of messages
	HTML::Lint::Constants	- constants used in the different modules

The "weblint" script is now pretty much a wrapper around the HTML::Lint module.
The following things aren't implemented in the weblint script at the moment:

	* checking of links (will be left for a separate program)
	* check URLs (using LWP)
	* Netscape & Microsoft extensions (do they have any these days?)

The distribution also includes weblint-cgi:

	* first cut at a web-based interface to weblint
	* requires LWP
	* doesn't have the overhead of external processes
	* this is a very early version

To install this you should just have to do:

	perl Makefile.PL
	make
	make test
	make install

The weblint-cgi script won't be installed.

Please let me know if any tests fail.

I'm using Perl 5.6.1.

Andy Lester
andy@petdance.com


Changes
Revision history HTML::Lint and Test::HTML::Lint.

2.00    Tue Sep 20 23:10:39 CDT 2005
        [CHANGES THAT COULD BREAK YOUR CODE]
        * I've changed the object structure.  HTML::Lint now has-a
          HTML::Parser, and no longer is-a HTML::Parser.

        * weblint-cgi and weblint-original are no longer distributed
          with HTML::Lint.

        * Now requires Perl 5.6.0.

        [FIXES]
        * Line numbering is now correct if you parse more than one file.

        * Changed t/*.*.t so that they only had one period in the
          filename. Apparently VMS doesn't like filenames with multiple
          dots.

1.30    December 3, 2004
        [ENHANCEMENTS]
        * Added handling for <HTML xmlns=... xml:lang=...>

        * Fixed <EMBED pluginspace=...>

        * Added handling of <EMBED quality=... and play=...>

1.28    January 27, 2004
        [ENHANCEMENTS]
        * Added support for the <EMBED> tag.

        [INTERNALS]
        * Added more POD coverage.

1.26    December 23, 2003
        [FIXES]
        * Fixed warnings if a lint error was found in the first column
          of the input file.  Thanks, Adam Monsen.

1.25    December 19, 2003
        [ENHANCEMENTS]
        * html_ok() clears the HTML::Lint object that gets passed
          in, if any.  Thanks to Cees Hek.

        [INTERNALS]
        * Ran everything through Devel::Cover to see what wasn't
          getting exercised.
        * Removed _element_stack() which was never used.
        * Error.pm's internals now eat their own dog food of methods,
          rather than accessing internals directly.
        * line(), column() and file() are no longer setters in
          HTML::Lint.  I sure hope this doesn't break anything for
          anyone.
        * Removed some explicit "return undef" stuff.

1.24    September 11, 2003
        [FIXES]
        * Removed the check for "input tags can have height &
          width", because they can't.

1.23    September 02, 2003
        [ENHANCEMENTS]
        * The types of errors to find may now be passed in the
          HTML::Lint constructor.
        * Can now export the error types from HTML::Lint::Error.

        [DOCUMENTATION]
        * Fixed a bug in Test::HTML::Lint docs. (Thanks to Leon
          Brocard and about 90 other people)

1.22    JUNE 11, 2003
        [ENHANCEMENTS]
        * Added some docs explaining html_ok()'s optional lint object.
        * Allowed the NAME attribute in the <IMG> tag.

        [FIXES]
        * Fixed the version check in t/99.pod.t

1.21    October 9, 2002
        [ENHANCEMENTS]
        * Added a test file for pod files
        * html_ok() now diag()s the test name before spewing the errors.
        * Made some docs a little more explicit. (Thanks to Ron Savage)
        * Added the check for missing <HTML>, <HEAD>, <TITLE> and <BODY>.
        * Added the check for nonrepeatable tags.

        [FIXES]
        * Ticket #1493: typo in HTML::Lint::HTML4 
          <ADDRESS> element in HTML4.pm was <ADDRESSS>.  Thanks to Dominic
          Mitchell for finding it.

1.20    August 22, 2002
        [ENHANCEMENTS]
        * Made it run under 5.5.3

1.13    August 5, 2002
        [ENHANCEMENTS]
        * Added text-use-entity for detecting control characters.

1.12    August 2, 2002
        [FIXES]
        * No longer squawks on html_ok( undef ).  It now specifically fails.

1.11    July 25, 2002
        [ENHANCEMENTS]
        * Added the concept of HTML::Lint::Error type, which is one of
          Structure, Helper or Fluff.
        * weblint now takes --[no]structure, --[no]helper, --[no]fluff.
        * Added the ability to overload html_ok()'s HTML::Lint object
          with one of your own.

        [FIXES]
        * Fixed line number goof in Test::HTML::Lint
        * Fixed line positions in weblint on files (URL reading was OK)

1.10    July 17, 2002
        * Added Test::HTML::Lint
        * Removed the "use 5.6.0" from everywhere
        * Included Neil Bowers' original weblint script, in the
          weblint-original/ directory.  Note that it does NOT get
          installed.

1.02    July 8, 2002
        * No functional changes.  It's all in the test suite.
        * Added a whole mess of *.t files.  They've been in CVS all along,
          but I forgot to put them in the MANIFEST.  Ooops.

1.01    July 3, 2002
        * Tests use isa_ok() for more stringent checking of return objects
        * Added <nobr>.  Ooopsie.

1.00    June 5, 2002
        * weblint has a --context option to show the line that the error
          occurs on.
        * Removed the original weblint log from this Changes file.
        * Fixed: Unknown tags wouldn't get put into the stack, so you
          would get two errors: One complaining that it didn't know the
          tag, and another that the closing tag didn't make sense.

0.94    May 31, 2002
        * Moved the %HTML::Lint::Error::errors hash to be a package
          variable, and not initialize it in an INIT block.

0.93    May 28, 2002    
        * weblint can now read URLs as well as files
        * Improved the docs in HTML::Lint::Error.

0.92    February 26, 2001
        * Removed the INIT block in HTML::Error in preparation for
          Apache::Lint, and so brian's weblint++ can use it.
        * errors() method respects wantarray (brian d foy)
        * file(), line(), column(), errcode() and errtext() all return
          blank instead of undef.

0.91    January 8, 2002
        * Fixed: Content description tags (<CITE>, <EM>, etc) were
          not seen as being valid tags.

0.90    July 8, 2001
        First version of the rewrite as a subclass of HTML::Parser

0.03    May 15, 2001
        * Fixed t/*.t problems

0.02    May 14, 2001
        * Packaged and bundled with new namespace
        * First upload to CPAN
        
0.01    April 20, 2001
        * original version; created by h2xs 1.21 with options
                -n HTML::Lint -X -A

lib/HTML/Lint/Error.pm
package HTML::Lint::Error;

use strict;

use base 'Exporter';
our @EXPORT = ();
our @EXPORT_OK = qw( STRUCTURE HELPER FLUFF );
our %EXPORT_TAGS = ( types => [@EXPORT_OK] );

our %errors;

=head1 NAME

HTML::Lint::Error - Error object for the Lint functionality

=head1 SYNOPSIS

See L<HTML::Lint> for all the gory details.

=head1 EXPORTS

None.  It's all object-based.

=head1 METHODS

Almost everything is an accessor.

=head1 Error types: C<STRUCTURE>, C<HELPER>, C<FLUFF>

Each error has a type.  Note that these roughly, but not exactly, go
from most severe to least severe.

=over 4

=item * C<STRUCTURE>

For problems that relate to the structural validity of the code.
Examples: Unclosed <TABLE> tags, incorrect values for attributes, and
repeated attributes.

=item * C<HELPER>

Helpers are notes that will help you with your HTML, or that will help
the browser render the code better or faster.  Example: Missing HEIGHT
and WIDTH attributes in an IMG tag.

=item * C<FLUFF>

Fluff is for items that don't hurt your page, but don't help it either.
This is usually something like an unknown attribute on a tag.

=back

=cut

use constant STRUCTURE  => 1;
use constant HELPER     => 2;
use constant FLUFF      => 3;

=head2 new()

Create an object.  It's not very exciting.

=cut

sub new {
    my $class = shift;

    my $file = shift;
    my $line = shift;
    my $column = shift;
    my $errcode = shift;
    my @errparms = @_;

    # Add an element that says what tag caused the error (B, TR, etc)
    # so that we can match 'em up down the road.
    my $self  = {
        _file => $file,
        _line => $line,
        _column => $column,
        _errcode => $errcode,
        _errtext => undef,
        _type => undef,
    };

    bless $self, $class;

    $self->_expand_error( $errcode, @errparms );

    return $self;
}

sub _expand_error {
    my $self = shift;

    my $errcode = shift;

    my $specs = $errors{$errcode};
    my $str;
    if ( $specs ) {
        ($str, $self->{_type}) = @$specs;
    }
    else {
        $str = "Unknown code: $errcode";
    }

    while ( @_ ) {
        my $var = shift;
        my $val = shift;
        $str =~ s/\$\{$var\}/$val/g;
    }

    $self->{_errtext} = $str;
}

=head2 is_type( $type1 [, $type2 ] )

Tells if any of I<$type1>, I<$type2>... match the error's type.
Returns the type that matched.

    if ( $err->is_type( HTML::Lint::Error::STRUCTURE ) ) {....

=cut

sub is_type {
    my $self = shift;

    for my $matcher ( @_ ) {
        return $matcher if $matcher eq $self->type;
    }

    return;
}

=head2 where()

Returns a formatted string that describes where in the file the
error has occurred.

For example,

    (14:23)

for line 14, column 23.

The terrible thing about this function is that it's both a plain
ol' formatting function as in

    my $str = where( 14, 23 );

AND it's an object method, as in:

    my $str = $error->where();

I don't know what I was thinking when I set it up this way, but
it's bad practice.

=cut

sub where {
    my $line;
    my $col;

    if ( not ref $_[0] ) {
        $line = shift;
        $col = shift;
    } else {
        my $self = shift;
        $line = $self->line;
        $col = $self->column;
    }
    $col ||= 0;
    return sprintf( "(%s:%s)", $line, $col + 1 );
}

=head2 as_string()

Returns a nicely-formatted string for printing out to stdout or some similar user thing.

=cut

sub as_string {
    my $self = shift;

    return sprintf( "%s %s %s", $self->file, $self->where, $self->errtext );
}

=head2 file()

Returns the filename of the error, as set by the caller.

=head2 line()

Returns the line number of the error.

=head2 column()

Returns the column number, starting from 0

=head2 errcode()

Returns the HTML::Lint error code.  Don't rely on this, because it will probably go away.

=head2 errtext()

Descriptive text of the error

=head2 type()

Type of the error

=cut

sub file        { my $self = shift; return $self->{_file}       || '' }
sub line        { my $self = shift; return $self->{_line}       || '' }
sub column      { my $self = shift; return $self->{_column}     || '' }
sub errcode     { my $self = shift; return $self->{_errcode}    || '' }
sub errtext     { my $self = shift; return $self->{_errtext}    || '' }
sub type        { my $self = shift; return $self->{_type}       || '' }


=head1 TODO

None, other than incorporating more errors, as driven by HTML::Lint.

=head1 LICENSE

This code may be distributed under the same terms as Perl itself.

Please note that these modules are not products of or supported by the
employers of the various contributors to the code.

=head1 AUTHOR

Andy Lester, C<andy at petdance.com>

=cut


# Errors that are commented out have not yet been implemented.

# Generic element stuff
%errors = (
    'elem-unknown' =>           ['Unknown element <${tag}>', STRUCTURE],
    'elem-unopened' =>          ['</${tag}> with no opening <${tag}>', STRUCTURE],
    'elem-unclosed' =>          ['<${tag}> at ${where} is never closed', STRUCTURE],
    'elem-empty-but-closed' =>  ['<${tag}> is not a container -- </${tag}> is not allowed', STRUCTURE],

    'elem-img-sizes-missing' => ['<IMG> tag has no HEIGHT and WIDTH attributes.', HELPER],
    'elem-img-alt-missing' =>   ['<IMG> does not have ALT text defined', HELPER],
    'elem-nonrepeatable' =>     ['<${tag}> is not repeatable, but already appeared at ${where}', STRUCTURE],

    'doc-tag-required' =>       ['<${tag}> tag is required', STRUCTURE],

    'attr-repeated' =>          ['${attr} attribute in <${tag}> is repeated', STRUCTURE],
    'attr-unknown' =>           ['Unknown attribute "${attr}" for tag <${tag}>', FLUFF],

    'text-use-entity' =>        ['Invalid character ${char} should be written as ${entity}', STRUCTURE],
);

1; # happy

__DATA__
Errors that haven't been done yet.

#elem-head-only                 <${tag}> can only appear in the <HEAD> element
#elem-non-head-element          <${tag}> cannot appear in the <HEAD> element
#elem-obsolete                  <${tag}> is obsolete
#elem-nested-element            <${tag}> cannot be nested -- one is already opened at ${where}
#elem-wrong-context             Illegal context for <${tag}> -- must appear in <${othertag}> tag.
#elem-heading-in-anchor         <A> should be inside <${tag}>, not <${tag}> inside <A>

#elem-head-missing              No <HEAD> element found
#elem-head-missing-title        No <TITLE> in <HEAD> element
#elem-img-sizes-incorrect       <IMG> tag's HEIGHT and WIDTH attributes are incorrect.  They should be ${correct}.
#attr-missing                   <${tag}> is missing a "${attr}" attribute

#comment-unclosed               Unclosed comment
#comment-markup                 Markup embedded in a comment can confuse some browsers

#text-literal-metacharacter     Metacharacter $char should be represented as "$otherchar"
#text-title-length              The HTML spec recommends that that <TITLE> be no more than 64 characters
#text-markup                    Tag <${tag}> found in the <TITLE>, which will not be rendered properly.

#elem-physical-markup           <${tag}> is physical font markup.  Use logical (such as <${othertag}>) instead.
#elem-leading-whitespace        <${tag}> should not have whitespace between "<" and "${tag}>"
#'must-follow' => [ ENABLED, MC_ERROR, '<$argv[0]> must immediately follow <$argv[1]>', ],
# 'empty-container' => [ ENABLED, MC_WARNING, 'empty container element <$argv[0]>.', ],
# 'directory-index' => [ ENABLED, MC_WARNING, 'directory $argv[0] does not have an index file ($argv[1])', ],
# 'attribute-delimiter' => [ ENABLED, MC_WARNING, 'use of \' for attribute value delimiter is not supported by all browsers (attribute $argv[0] of tag $argv[1])', ],
# 'container-whitespace' => [ DISABLED, MC_WARNING, '$argv[0] whitespace in content of container element $argv[1]', ],
# 'bad-text-context' => [ ENABLED, MC_ERROR, 'illegal context, <$argv[0]>, for text; should be in $argv[1].', ],
# 'attribute-format' => [ ENABLED, MC_ERROR, 'illegal value for $argv[0] attribute of $argv[1] ($argv[2])', ],
# 'quote-attribute-value' => [ ENABLED, MC_ERROR, 'value for attribute $argv[0] ($argv[1]) of element $argv[2] should be quoted (i.e. $argv[0]="$argv[1]")', ],
# 'meta-in-pre' => [ ENABLED, MC_ERROR, 'you should use "$argv[0]" in place of "$argv[1]", even in a PRE element.', ],
#  'implied-element' => [ ENABLED, MC_WARNING, 'saw <$argv[0]> element, but no <$argv[1]> element', ],
#  'button-usemap' => [ ENABLED, MC_ERROR, 'illegal to associate an image map with IMG inside a BUTTON', ],


lib/HTML/Lint.pm
package HTML::Lint;

=head1 NAME

HTML::Lint - check for HTML errors in a string or file

=head1 VERSION

Version 2.00

=cut

our $VERSION = '2.00';

=head1 SYNOPSIS

    my $lint = HTML::Lint->new;
    $lint->only_types( HTML::Lint::STRUCTURE );

    $lint->parse( $data );
    $lint->parse_file( $filename );

    my $error_count = $lint->errors;

    foreach my $error ( $lint->errors ) {
        print $error->as_string, "\n";
    }

HTML::Lint also comes with a wrapper program called F<weblint> that handles
linting from the command line:

    $ weblint http://www.cnn.com/
    http://www.cnn.com/ (395:83) <IMG> tag has no HEIGHT and WIDTH attributes.
    http://www.cnn.com/ (395:83) <IMG> does not have ALT text defined
    http://www.cnn.com/ (396:217) Unknown element <nobr>
    http://www.cnn.com/ (396:241) </nobr> with no opening <nobr>
    http://www.cnn.com/ (842:7) target attribute in <a> is repeated

And finally, you can also get L<Apache::HTML::Lint> that passes any
mod_perl-generated code through HTML::Lint and get it dumped into your
Apache F<error_log>.

    [Mon Jun  3 14:03:31 2002] [warn] /foo.pl (1:45) </p> with no opening <p>
    [Mon Jun  3 14:03:31 2002] [warn] /foo.pl (1:49) Unknown element <gronk>
    [Mon Jun  3 14:03:31 2002] [warn] /foo.pl (1:56) Unknown attribute "x" for tag <table>

=cut

use strict;
use HTML::Lint::Error;

=head1 METHODS

NOTE: Some of these methods mirror L<HTML::Parser>'s methods, but HTML::Lint
is not a subclass of HTML::Parser.

=head2 new()

Create an HTML::Lint object, which inherits from HTML::Parser.
You may pass the types of errors you want to check for in the
C<only_types> parm.

    my $lint = HTML::Lint->new( only_types => HTML::Lint::Error::STRUCTURE );

If you want more than one, you must pass an arrayref:

    my $lint = HTML::Lint->new( 
        only_types => [HTML::Lint::Error::STRUCTURE, HTML::Lint::Error::FLUFF] );

=cut

sub new {
    my $class = shift;
    my %args = @_;

    my $self = {
        _errors => [],
        _types => [],
    };
    $self->{_parser} = HTML::Lint::Parser->new( sub { $self->gripe( @_ ) } );
    bless $self, $class;

    if ( my $only = $args{only_types} ) {
        $self->only_types( ref $only eq "ARRAY" ? @$only : $only );
        delete $args{only_types};
    }

    warn "Unknown argument $_\n" for keys %args;

    return $self;
}

=head2 $lint->parse( $chunk )

=head2 $lint->parse( $code_ref )

Passes in a chunk of HTML to be linted, either as a piece of text,
or a code reference.
See L<HTML::Parser>'s C<parse_file> method for details.

=cut

sub parse {
    my $self = shift;
    return $self->_parser->parse( @_ );
}

=head2 $lint->parse_file( $file )

Analyzes HTML directly from a file. The C<$file> argument can be a filename,
an open file handle, or a reference to an open file handle.
See L<HTML::Parser>'s C<parse_file> method for details.

=cut

sub parse_file {
    my $self = shift;
    return $self->_parser->parse_file( @_ );
}

=head2 $lint->eof

Signals the end of a block of text getting passed in.  This must be
called to make sure that all parsing is complete before looking at errors.

Any parameters (and there shouldn't be any) are passed through to
HTML::Parser's eof() method.

=cut

sub eof {
    my $self = shift;

    my $rc;
    my $parser = $self->_parser;
    if ( $parser ) {
        $rc = $self->_parser->eof(@_);
        delete $self->{_parser};
    }

    return $rc;
}

=head2 $lint->errors()

In list context, C<errors> returns all of the errors found in the
parsed text.  Each error is an object of the type L<HTML::Lint::Error>.

In scalar context, it returns the number of errors found.

=cut

sub errors {
    my $self = shift;

    if ( wantarray ) {
        return @{$self->{_errors}};
    }
    else {
        return scalar @{$self->{_errors}};
    }
}

=head2 $lint->clear_errors()

Clears the list of errors, in case you want to print and clear, print and clear.

=cut

sub clear_errors {
    my $self = shift;

    $self->{_errors} = [];
}

=head2 $lint->only_types( $type1[, $type2...] )

Specifies to only want errors of a certain type.

    $lint->only_types( HTML::Lint::Error::STRUCTURE );

Calling this without parameters makes the object return all possible
errors.

The error types are C<STRUCTURE>, C<HELPER> and C<FLUFF>.
See L<HTML::Lint::Error> for details on these types.

=cut

sub only_types {
    my $self = shift;

    $self->{_types} = [@_];
}

=head2 $lint->gripe( $errcode, [$key1=>$val1, ...] )

Adds an error message, in the form of an L<HTML::Lint::Error> object,
to the list of error messages for the current object.  The file,
line and column are automatically passed to the L<HTML::Lint::Error>
constructor, as well as whatever other key value pairs are passed.

For example:

    $lint->gripe( 'attr-repeated', tag => $tag, attr => $attr );

Usually, the user of the object won't call this directly, but just
in case, here you go.

=cut

sub gripe {
    my $self = shift;

    my $parser = $self->_parser;

    my $error = new HTML::Lint::Error(
        $self->{_file}, $parser->{_line}, $parser->{_column}, @_ );

    my @keeps = @{$self->{_types}};
    if ( !@keeps || $error->is_type(@keeps) ) {
        push( @{$self->{_errors}}, $error );
    }
}

=head2 $lint->newfile( $filename )

Call C<newfile()> whenever you switch to another file in a batch of 
linting.  Otherwise, the object thinks everything is from the same file.
Note that the list of errors is NOT cleared.

Note that I<$filename> does NOT need to match what's put into parse()
or parse_file().  It can be a description, a URL, or whatever.

=cut

sub newfile {
    my $self = shift;
    my $file = shift;

    delete $self->{_parser};
    $self->{_parser} = HTML::Lint::Parser->new( sub { $self->gripe( @_ ) } );
    $self->{_file} = $file;
    $self->{_line} = 0;
    $self->{_column} = 0;
    $self->{_first_seen} = {};

    return $self->{_file};
} # newfile

sub _parser {
    my $self = shift;

    return $self->{_parser};
}

=pod

HTML::Lint::Parser is a class only for this module.

=cut

package HTML::Lint::Parser;

use HTML::Parser 3.20;
use HTML::Tagset 3.03;

use HTML::Lint::HTML4 qw( %isKnownAttribute %isRequired %isNonrepeatable %isObsolete );
use HTML::Entities qw( %char2entity );

our @ISA = qw( HTML::Parser );

sub new {
    my $class = shift;
    my $gripe = shift;

    my $self =
        HTML::Parser->new(
            api_version => 3,
            start_document_h    => [ \&_start_document, 'self' ],
            end_document_h      => [ \&_end_document,   'self,line,column' ],
            start_h             => [ \&_start,          'self,tagname,line,column,@attr' ],
            end_h               => [ \&_end,            'self,tagname,line,column,@attr' ],
            text_h              => [ \&_text,           'self,text' ],
            strict_names => 1,
        );
    bless $self, $class;

    $self->{_gripe} = $gripe;
    $self->{_stack} = [];

    return $self;
}

sub gripe {
    my $self = shift;

    return $self->{_gripe}->( @_ );
}

sub _start_document {
    my $self = shift;
}

sub _end_document {
    my ($self,$line,$column) = @_;

    for my $tag ( keys %isRequired ) {
        if ( !$self->{_first_seen}->{$tag} ) {
            $self->gripe( 'doc-tag-required', tag => $tag );
        }
    }
}

sub _start {
    my ($self,$tag,$line,$column,@attr) = @_;

    $self->{_line} = $line;
    $self->{_column} = $column;

    my $validattr = $isKnownAttribute{ $tag };
    if ( $validattr ) {
        my %seen;
        my $i = 0;
        while ( $i < @attr ) {
            my ($attr,$val) = @attr[$i++,$i++];
            if ( $seen{$attr}++ ) {
                $self->gripe( 'attr-repeated', tag => $tag, attr => $attr );
            }

            if ( $validattr && ( !$validattr->{$attr} ) ) {
                $self->gripe( 'attr-unknown', tag => $tag, attr => $attr );
            }
        } # while attribs
    }
    else {
        $self->gripe( 'elem-unknown', tag => $tag );
    }
    $self->_element_push( $tag ) unless $HTML::Tagset::emptyElement{ $tag };

    if ( my $where = $self->{_first_seen}{$tag} ) {
        if ( $isNonrepeatable{$tag} ) {
            $self->gripe( 'elem-nonrepeatable',
                            tag => $tag,
                            where => HTML::Lint::Error::where(@$where)
                        );
        }
    }
    else {
        $self->{_first_seen}{$tag} = [$line,$column];
    }

    # Call any other overloaded func
    my $tagfunc = "_start_$tag";
    if ( $self->can($tagfunc) ) {
        $self->$tagfunc( $tag, @attr );
    }
}

sub _text {
    my ($self,$text) = @_;

    while ( $text =~ /([^\x09\x0A\x0D -~])/g ) {
        my $bad = $1;
        $self->gripe(
            'text-use-entity', 
                char => sprintf( '\x%02lX', ord($bad) ),
                entity => $char2entity{ $bad },
        );
    }
}

sub _end {
    my ($self,$tag,$line,$column,@attr) = @_;

    $self->{_line} = $line;
    $self->{_column} = $column;

    if ( $HTML::Tagset::emptyElement{ $tag } ) {
        $self->gripe( 'elem-empty-but-closed', tag => $tag );
    }
    else {
        if ( $self->_in_context($tag) ) {
            my @leftovers = $self->_element_pop_back_to($tag);
            for ( @leftovers ) {
                my ($tag,$line,$col) = @$_;
                $self->gripe( 'elem-unclosed', tag => $tag, 
                        where => HTML::Lint::Error::where($line,$col) )
                        unless $HTML::Tagset::optionalEndTag{$tag};
            } # for
        }
        else {
            $self->gripe( 'elem-unopened', tag => $tag );
        }
    } # is empty element

    # Call any other overloaded func
    my $tagfunc = "_end_$tag";
    if ( $self->can($tagfunc) ) {
        $self->$tagfunc( $tag, $line );
    }
}

sub _element_push {
    my $self = shift;
    for ( @_ ) {
        push( @{$self->{_stack}}, [$_,$self->{_line},$self->{_column}] );
    } # while
}

sub _find_tag_in_stack {
    my $self = shift;
    my $tag = shift;
    my $stack = $self->{_stack};

    my $offset = @$stack - 1;
    while ( $offset >= 0 ) {
        if ( $stack->[$offset][0] eq $tag ) {
            return $offset;
        }
        --$offset;
    } # while

    return;
}

sub _element_pop_back_to {
    my $self = shift;
    my $tag = shift;

    my $offset = $self->_find_tag_in_stack($tag) or return;

    my @leftovers = splice( @{$self->{_stack}}, $offset + 1 );
    pop @{$self->{_stack}};

    return @leftovers;
}

sub _in_context {
    my $self = shift;
    my $tag = shift;

    my $offset = $self->_find_tag_in_stack($tag);
    return defined $offset;
}

# Overridden tag-specific stuff
sub _start_img {
    my ($self,$tag,%attr) = @_;

    my ($h,$w) = @attr{qw( height width )};
    if ( defined $h && defined $w ) {
        # Check sizes
    }
    else {
        $self->gripe( "elem-img-sizes-missing" );
    }
    if ( not defined $attr{alt} ) {
        $self->gripe( "elem-img-alt-missing" );
    }
}

=head1 BUGS, WISHES AND CORRESPONDENCE

Please feel free to email me at andy@petdance.com.  I'm glad to help as
best I can, and I'm always interested in bugs, suggestions and patches.

Please report any bugs or feature requests to
C<< <bug-html-lint at rt.cpan.org> >>, or through the web interface at
L<http://rt.cpan.org>.  I will be notified, and then you'll automatically
be notified of progress on your bug as I make changes.

=head1 TODO

=over 4

=item * Check for attributes that require values

=item * <TABLE>s that have no rows.

=item * Form fields that aren't in a FORM

=item * Check for valid entities, and that they end with semicolons

=item * DIVs with nothing in them.

=item * HEIGHT= that have percents in them.

=item * Check for goofy stuff like:

    <b><li></b><b>Hello Reader - Spanish Level 1 (K-3)</b>

=back

=head1 LICENSE

Copyright 2005 Andy Lester, All Rights Reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

Please note that these modules are not products of or supported by the
employers of the various contributors to the code.

=head1 AUTHOR

Andy Lester, andy at petdance.com

=cut

